File: evaluator/problem_suites/polyglot/TEST_RUNNER.js
/**
 * Polyglot Test Runner - JavaScript Version
 * 
 * This script runs inside the Docker sandbox to execute JavaScript tests
 * for the polyglot problem suite. It provides a Jest-like testing environment
 * and runs tests defined in /sandbox/repo/tests.js.
 * 
 * Execution Flow:
 * 1. Sets up Jest-like globals (describe, test, expect, beforeEach, afterEach)
 * 2. Loads and executes tests.js from the repository
 * 3. Collects test results as tests execute
 * 4. Writes structured results to output.json
 * 
 * Expected File Layout:
 *     /sandbox/repo/
 *         ├── main.js    # Code to test
 *         └── tests.js   # Jest-style test definitions
 * 
 * Jest-like API Provided:
 *     - describe(name, fn): Define a test suite
 *     - test(name, fn): Define a test case
 *     - expect(value): Assertion interface
 *     - beforeEach(fn): Setup hook
 *     - afterEach(fn): Teardown hook
 * 
 * Output Format:
 *     [
 *         {"name": "test description", "category": "default", "status": "pass|fail|skip"},
 *         ...
 *     ]
 */

const fs = require('fs');
const path = require('path');
const {expect} = require('expect');



// Path to the code repository inside the sandbox
const repoPath = "/sandbox/repo";



function runTests() {
    /** Discover and run all tests in the repository. */
    
    // Array to collect test results as they execute
    global.testResults = [];



    // ============================================
    // Jest API Emulator
    // ============================================
    // Provides Jest-like globals for test files to use
    // This allows tests to be written in standard Jest style
    
    global._beforeEachCallbacks = [];
    global.beforeEach = function(callback) {
        /** Register a setup function to run before each test. */
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] beforeEach()`);
        global._beforeEachCallbacks.push(callback);
    };

    global._afterEachCallbacks = [];
    global.afterEach = function(callback) {
        /** Register a teardown function to run after each test. */
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] afterEach()`);
        global._afterEachCallbacks.push(callback);
    };

    global.describe = function(description, callback) {
        /** Define a test suite (group of related tests). */
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] describe(): "${description}"`);
        callback();
    };
    
    global.xdescribe = function(description, callback) {
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] xdescribe(): "${description}"`);
        describe(description, callback);
    };

    global.test = function(description, callback) {
        /** Define and immediately execute a test case.
         * 
         * Runs beforeEach hooks, the test, then afterEach hooks.
         * Captures success/failure and records the result.
         */
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] Test: "${description}"...`);
        try {
            // Run setup hooks
            for (const beforeEachCallback of global._beforeEachCallbacks)
                beforeEachCallback();
            
            // Execute the test
            callback();
            
            // Run teardown hooks
            for (const afterEachCallback of global._afterEachCallbacks)
                afterEachCallback();
            
            console.log(`[POLYGLOT_TEST_RUNNER] [JEST] Test passed: "${description}"`);
            global.testResults.push({
                "name": description,
                "category": "default",
                "status": "pass"
            });
        } catch (e) {
            // Test failed - capture the error
            console.log(`[POLYGLOT_TEST_RUNNER] [JEST] Test failed: "${description}"`);
            console.log(e);
            global.testResults.push({
                "name": description,
                "category": "default",
                "status": "fail"
            });
        }
    };

    global.xtest = function(description, callback) {
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] xtest(): "${description}"`);
        test(description, callback);
    };

    global.test.skip = function(description, callback) {
        console.log(`[POLYGLOT_TEST_RUNNER] [JEST] test.skip(): "${description}"`);
    };

    // Make expect available for assertions in tests
    global.expect = expect;



    // Load and execute the test file
    // This will trigger all the test() calls which run immediately
    console.log("[POLYGLOT_TEST_RUNNER] Loading tests.js");
    require(path.join(repoPath, "tests.js"));
    console.log("[POLYGLOT_TEST_RUNNER] Loaded tests.js");
    
    return global.testResults;
}



function main() {
    /** Main entry point for test execution inside the sandbox. */
    console.log("[POLYGLOT_TEST_RUNNER] Entered main()");
    
    try {
        // Run all tests and collect results
        const testResults = runTests();

        // Print results for debugging (visible in container logs)
        console.log("[POLYGLOT_TEST_RUNNER] Test results:");
        console.log(JSON.stringify(testResults, null, 2));
        
        const testsPassed = testResults.filter(test => test["status"] === "pass").length;
        const testsFailed = testResults.filter(test => test["status"] === "fail").length;
        const testsSkipped = testResults.filter(test => test["status"] === "skip").length;
        
        console.log(`[POLYGLOT_TEST_RUNNER] Test summary: ${testsPassed} passed, ${testsFailed} failed, ${testsSkipped} skipped`);

        console.log("[POLYGLOT_TEST_RUNNER] Writing output.json");
        fs.writeFileSync("/sandbox/output.json", JSON.stringify({
            "success": true,
            "output": testResults
        }, null, 2));
        console.log("[POLYGLOT_TEST_RUNNER] Wrote output.json");

    } catch (e) {
        console.log("[POLYGLOT_TEST_RUNNER] Exception:");
        console.log(e.stack || e);
        
        const output = {
            "success": false,
            "error": String(e),
            "traceback": e.stack || ""
        };
        
        try {
            console.log("[POLYGLOT_TEST_RUNNER] Writing output.json");
            fs.writeFileSync("/sandbox/output.json", JSON.stringify(output, null, 2));
            console.log("[POLYGLOT_TEST_RUNNER] Wrote output.json");
        } catch {
            console.log("[POLYGLOT_TEST_RUNNER] Failed to write output.json");
        }
    }

    console.log("[POLYGLOT_TEST_RUNNER] Exiting main()");
}



if (require.main === module) {
    main();
}

File: evaluator/problem_suites/polyglot/TEST_RUNNER.py
"""Polyglot Test Runner - Python Version

This script runs inside the Docker sandbox to execute Python unit tests
for the polyglot problem suite. It discovers and runs unittest-based tests
in the /sandbox/repo directory.

Execution Flow:
1. Loads the main.py module (the code being tested)
2. Loads the tests.py module (unittest test cases)
3. Discovers all test methods in the test class
4. Runs each test and collects results
5. Writes structured results to output.json

Expected File Layout:
    /sandbox/repo/
        ├── main.py    # Code to test (loaded as module)
        └── tests.py   # Unittest test cases

Output Format:
    [
        {"name": "test_method_1", "category": "default", "status": "pass|fail|skip"},
        ...
    ]

Note:
    - Tests run sequentially, not in parallel
    - All tests are discovered from a single TestCase subclass
    - Test execution continues even after failures
"""

import os
import sys
import json
import unittest
import traceback
import importlib.util



# Add the repository directory to Python path so imports work
repo_path = "/sandbox/repo"
sys.path.insert(0, repo_path)



def run_tests():
    """Discover and run all tests in the repository.
    
    Returns:
        List of test result dictionaries with name, category, and status
    """
    # Load the main module (the code being tested)
    # This makes the implementation available to the tests
    print("[POLYGLOT_TEST_RUNNER] Loading main.py")
    main_spec = importlib.util.spec_from_file_location("main", "/sandbox/repo/main.py")
    main_module = importlib.util.module_from_spec(main_spec)
    main_spec.loader.exec_module(main_module)
    print("[POLYGLOT_TEST_RUNNER] Loaded main.py")
    
    # Load tests module
    print("[POLYGLOT_TEST_RUNNER] Loading tests.py")
    tests_spec = importlib.util.spec_from_file_location("tests", "/sandbox/repo/tests.py")
    tests_module = importlib.util.module_from_spec(tests_spec)
    tests_spec.loader.exec_module(tests_module)
    print("[POLYGLOT_TEST_RUNNER] Loaded tests.py")
    
    test_class = None
    for name in dir(tests_module):
        obj = getattr(tests_module, name)
        if (isinstance(obj, type) and issubclass(obj, unittest.TestCase) and obj is not unittest.TestCase):
            test_class = obj
            break
    
    if not test_class:
        raise Exception("No test class found in tests.py")
    
    print(f"[POLYGLOT_TEST_RUNNER] Found test class: {test_class.__name__}")
    
    test_methods = [method for method in dir(test_class) if method.startswith("test_")]
    print(f"[POLYGLOT_TEST_RUNNER] Found {len(test_methods)} test methods")
    


    test_results = []
    for method_name in test_methods:
        test_results.append({"name": method_name, "category": "default", "status": "skip"})
    
    total_tests = len(test_results)
    test_instance = test_class()
    
    for test_index, test_result in enumerate(test_results, 1):
        method_name = test_result["name"]
        
        try:
            print(f"[POLYGLOT_TEST_RUNNER] [{test_index}/{total_tests}] Running {method_name}...")
            method = getattr(test_instance, method_name)
            method()
            print(f"[POLYGLOT_TEST_RUNNER] {method_name}: PASSED")
            test_result["status"] = "pass"
        except Exception as e:
            print(f"[POLYGLOT_TEST_RUNNER] {method_name}: FAILED - {e}")
            test_result["status"] = "fail"
            # break
    
    return test_results



def main():
    """Main entry point for test execution inside the sandbox."""
    print("[POLYGLOT_TEST_RUNNER] Entered main()")
    
    try:
        # Run all tests and collect results
        test_results = run_tests()

        # Print results for debugging (visible in container logs)
        print("[POLYGLOT_TEST_RUNNER] Test results:")
        print(json.dumps(test_results, indent=2))
        
        tests_passed = sum(1 for test in test_results if test["status"] == "pass")
        tests_failed = sum(1 for test in test_results if test["status"] == "fail")
        tests_skipped = sum(1 for test in test_results if test["status"] == "skip")
        
        print(f"[POLYGLOT_TEST_RUNNER] Test summary: {tests_passed} passed, {tests_failed} failed, {tests_skipped} skipped")

        print("[POLYGLOT_TEST_RUNNER] Writing output.json")
        with open("/sandbox/output.json", "w") as f:
            json.dump({
                "success": True,
                "output": test_results
            }, f, indent=2)
        print("[POLYGLOT_TEST_RUNNER] Wrote output.json")

    except Exception as e:
        print("[POLYGLOT_TEST_RUNNER] Exception:")
        traceback.print_exc(file=sys.stdout)
        
        output = {
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }
        
        try:
            print("[POLYGLOT_TEST_RUNNER] Writing output.json")
            with open("/sandbox/output.json", "w") as f:
                json.dump(output, f, indent=2)
            print("[POLYGLOT_TEST_RUNNER] Wrote output.json")
        except:
            print("[POLYGLOT_TEST_RUNNER] Failed to write output.json")
            pass

    print("[POLYGLOT_TEST_RUNNER] Exiting main()")



if __name__ == "__main__":
    main()